Behaviors UI

michalczaplinski opened the pull request:
## What?
Adds a simple UI element to add specific behaviors to a block and support for them to `theme.json` & `block.json`. It is only concerned with the built-in "LIGHTBOX" behavior that is worked on in #49621.

The behavior can be defined in the following way:

```js
// theme.json

{  
  // We can toggle the Behaviors UI on/off for specific blocks. 
  // Currently it will only be available for the core/image block.
  "settings": {
    "blocks": {
      "core/image": {
        "behaviorsUIEnabled": true
      }
    }
  },
  // Here we can define the actual behaviors that are available in the theme.
  "behaviors": {
    "lightbox": true,
   }
}
```

## Why?
We need a way for users to add specific Behaviors to their blocks. We also need them to be able to opt into them in the `theme.json` file.

## How?
- Using the editor.BlockEdit filter to add the relevant UI like:

https://github.com/WordPress/gutenberg/blob/77288e42e762c30f3599751c1810339e8aca7f69/packages/block-editor/src/hooks/behaviors.js#L83-L88

- Updating the core `theme.json` file so that the core Image block uses the LIGHTBOX behavior by default.


## Testing Instructions
1. Insert an Image block in a post
2. Go to the "Advanced" tab in the Inspector
3. You can select the LIGHTBOX behavior or choose "None" in the dropdown

## Screenshots

https://github.com/WordPress/gutenberg/assets/5417266/50bf3b36-6939-4f1d-947c-f90b0e334f34

michalczaplinski committed:
Add behaviors to the core theme.json

michalczaplinski committed:
Add behaviors to json schemas for theme.json

gziolo commented on schemas/json/theme.json:
```diff
@@ -796,7 +797,28 @@
+									"type": "object",
+									"properties": {
+										"lightbox": {
+											"description": "Show a lightbox for the image.\nGutenberg plugin required.",
```
The wording is inconsistent in the entire `theme.json` schema, so it's difficult to understand the intent for all settings. However, I believe it's all about deciding whether a UI control should be available to users. It's still up to the block's author to decide what to do about it, so the lightbox might still be enabled - it's just the user can't control it.

luisherranz replied:
> so the lightbox might still be enabled - it's just the user can't control it

Oh, I think we didn't know this bit.

cc: @santosguillamot

SantosGuillamot replied:
Me neither! Thanks for sharing 🙂

ajlende commented on schemas/json/theme.json:
```diff
@@ -796,7 +797,28 @@
+									"type": "object",
+									"properties": {
+										"lightbox": {
+											"description": "Show a lightbox for the image.\nGutenberg plugin required.",
```
> Gutenberg plugin required.

Is this hidden behind an experiment flag? If yes, the schema should not be added at all. If no, this part should be removed as you are editing the plugin version of the schema. We don't do any processing of the schemas for release so it will become outdated when we tag the wp/6.3 version.

michalczaplinski replied:
Makes sense! I've removed the the schema now 👍

michalczaplinski committed:
Add a behaviors panel

michalczaplinski committed:
Remove the changes to theme.json schema

michalczaplinski committed:
Add behaviors to the VALID_SETTINGS in class-wp-theme-json-gutenberg.php

michalczaplinski committed:
Add the first (still broken) version of the lightbox settings

michalczaplinski commented on packages/block-editor/src/hooks/behaviors.js:
```diff
@@ -0,0 +1,70 @@
+export const withBehaviors = createHigherOrderComponent( ( BlockEdit ) => {
+	return ( props ) => {
+		const { behaviors: blockBehaviors } = props.attributes;
+		const settings = select( blockEditorStore ).getSettings();
+		const themeBehaviors =
+			settings?.__experimentalFeatures?.blocks?.[ 'core/image' ]
+				.behaviors;
```
If anyone knows a more idiomatic way to access the block-level settings from the
`theme.json`, I'm all ears!

cbravobernal replied:
Not more idiomatic, but I guess using props.name instead of the core block hard code name would be more scalable. Something like:

```js
const themeBehaviors = settings?.__experimentalFeatures?.blocks?.[ props.name ]?.behaviors;
```

michalczaplinski replied:
Makes sense 🙂 

Even though, for now, we're only enabling the Behaviors on the Image block, we should still change it so that we don't have to update this line in the future.

michalczaplinski replied:
Done dd70de9b14337e164dc23f5b076889dbf468c5fd

michalczaplinski commented:
I've created a UI to enable the Behaviors in the "Advanced" panel of the block InspectorControls. It's still still WIP and completely broken at the moment, but it's a skeleton to build off.

I'm using the `editor.BlockEdit` filter to add the relevant UI like:

https://github.com/WordPress/gutenberg/blob/5afa6d362a33ebb10ddb6e6437034ea86b547302/packages/block-editor/src/hooks/behaviors.js#L66-L70

Just using `.getSettings()` seems enough to make the system use the value from the theme's `theme.json` and fall back on the core `theme.json` file if not present in the theme's file. The next step should be to ensure that the `block.json`'s setting for behaviors is respected.

michalczaplinski committed:
WIP: Added a SelectControl for the behaviors

cbravobernal commented:
> I've created a UI to enable the Behaviors in the "Advanced" panel of the block InspectorControls. It's still still WIP and completely broken at the moment, but it's a skeleton to build off.
> 
> I'm using the `editor.BlockEdit` filter to add the relevant UI like:
> 
> https://github.com/WordPress/gutenberg/blob/5afa6d362a33ebb10ddb6e6437034ea86b547302/packages/block-editor/src/hooks/behaviors.js#L66-L70
> 
> Just using `.getSettings()` seems enough to make the system use the value from the theme's `theme.json` and fall back on the core `theme.json` file if not present in the theme's file. The next step should be to ensure that the `block.json`'s setting for behaviors is respected.

Should we use the behaviors inside block supports? Imagine the block.json file of image.
```json
"supports": {
		"__experimentalBorder": {
			"color": true,
			"radius": true,
			"width": true,
			"__experimentalSkipSerialization": true,
			"__experimentalDefaultControls": {
				"color": true,
				"radius": true,
				"width": true
			}
		},
		"__experimentalBehaviors": {
			"lightbox": false
		}
	},
```
That way we could conditionally check if the block supports the behaviors, which ones are supported and show the UI so the user can select which ones to apply. Something like this:
```js
// Check if the block has behaviors.
const blockBehaviors = select( blocksStore ).getBlockSupport(
  props.name,
  '__experimentalBehaviors'
 );
```

michalczaplinski committed:
Format PHP

michalczaplinski committed:
Format correctly again

michalczaplinski commented:
> Should we use the behaviors inside block supports? Imagine the block.json file of image.

I haven't considered that, but it might make sense! 

This way, individual blocks could enable/disable the Behaviors UI using the `supports` as you mentioned. Then the individual behaviors can be enabled/disabled in the `block.json` of a block.

I've added the outline of the technical details of the implementation (previously created by @SantosGuillamot), which briefly mentions the block supports in https://github.com/WordPress/gutenberg/issues/50029#issuecomment-1540858172 . Let's discuss this over there in the issue!

michalczaplinski committed:
Use the props.name when getting the behaviors

cbravobernal committed:
Add initial e2e tests

cbravobernal commented:
I added a couple e2e tests. Feel free to update them if needed!

michalczaplinski committed:
Update the withBehaviors description

michalczaplinski committed:
Cleaned up behaviors.js

michalczaplinski commented:
The tests look great but I have some trouble running them locally @c4rl0sbr4v0 ! 

```
  Error: The theme "behaviors" is not installed
```

I've tried changing that to `gutenberg-test-themes/behaviors` to match how it's defined in the `style-variations.spec.js`: 

https://github.com/WordPress/gutenberg/blob/cddc3d4e62f15c9686358a69ba417e14a7f86eac/test/e2e/specs/site-editor/style-variations.spec.js#L14-L16

but that does not help either. Can you run them locally?

michalczaplinski commented:
I've pushed some more changes and I believe the implementation is close to what we want. There are still a few things left:

- [x] Figure out the right way to get the value from the top-level property of the `theme.json`
   https://github.com/WordPress/gutenberg/blob/7f71175f903c7c8910e4912cd592fdaced6cfcdb/packages/block-editor/src/hooks/behaviors.js#L45-L49
- [x] Fix the currently failing e2e tests
- [x] Add additional e2e tests:
  - Check that the block's value for the behavior takes precedence over the theme's value.
  - Check that if behavior is not set in the block, the value from the `theme.json` is used.
  - Check that the Behaviors **UI** can be enabled/disabled based on the value in `theme.json`
- [x] Fix the linting CI checks that are failing right now
- [x] Make sure we consciously use `__experimentalFeatures`. It seems to be built for Global Styles, so I'm not 100% sure if our usage is correct. We use it in:
  https://github.com/WordPress/gutenberg/blob/7f71175f903c7c8910e4912cd592fdaced6cfcdb/packages/block-editor/src/hooks/behaviors.js#L37
  https://github.com/WordPress/gutenberg/blob/7f71175f903c7c8910e4912cd592fdaced6cfcdb/packages/block-editor/src/hooks/behaviors.js#L49
- [ ] Coordinate with @artemiomorales about integrating with https://github.com/WordPress/gutenberg/pull/50373. I think that the current PR should be merged first and afterwards in #50373 we can make the following small adjustments:
  - Rename the `enableLightbox` to just `lightbox`.
  - Remove the changes to the `block.json`: https://github.com/WordPress/gutenberg/pull/50373/files#diff-4e637f96eb9e465696d35d73a41e6a694c7718f15fb805b49d49931262dfc53a
  - Remove the custom InspectorControls in https://github.com/WordPress/gutenberg/pull/50373/files#diff-a14b61c874095649b236b78413e805b4541277b24eb71ce350a17d9f2b6396b1R464-R481

cbravobernal committed:
Update e2e tests to check visibility

cbravobernal commented:
I updated the branch with trunk and did
- [x] Fix the currently failing e2e tests

artemiomorales commented:
Hey @michalczaplinski, it's great to see this coming along!

I looked at [your video](https://github.com/WordPress/gutenberg/issues/50029#issuecomment-1544479640) on the Outline for Block Behavior issue explaining your ideas on the UX and have two questions:

1.  Do you think we should add the toggle into the UI right now?

2. How do you feel about adding the ability to enable or disable the lightboxes by default?

I'm wondering what would be the best way to ship this out to users. What we have right now will make the lightbox available to folks, but if our intent is to generate ideas and discussion around behaviors, maybe we can consider adding these two points.

cbravobernal commented:
> The tests look great but I have some trouble running them locally @c4rl0sbr4v0 !
> 
> ```
>   Error: The theme "behaviors" is not installed
> ```
> 
> I've tried changing that to `gutenberg-test-themes/behaviors` to match how it's defined in the `style-variations.spec.js`:
> 
> https://github.com/WordPress/gutenberg/blob/cddc3d4e62f15c9686358a69ba417e14a7f86eac/test/e2e/specs/site-editor/style-variations.spec.js#L14-L16
> 
> but that does not help either. Can you run them locally?

Yes! I run 
`npm run test:e2e:playwright test/e2e/specs/editor/various/behaviors.spec.js` and I get:

![Screenshot 2023-05-12 at 14 16 22](https://github.com/WordPress/gutenberg/assets/37012961/992a69b3-6bcf-45b5-abe5-ef3539998c2c)

I've created the theme like the style-variations you mention. Let's see if it also fails in CI.

cbravobernal committed:
Update core-blocks doc with behaviors attribute

cbravobernal committed:
Update fixtures with new lightbox attribute

cbravobernal commented:
The CI is now green 🎉

michalczaplinski commented:
> Yes! I run
npm run test:e2e:playwright test/e2e/specs/editor/various/behaviors.spec.js and I get:

Yeah, not sure what happened there. I guess something got messed up in my local docker setup. I ran `npx wp-env clean all` and that sorted it out.

michalczaplinski commented:
@artemiomorales 

> Do you think we should add the toggle into the UI right now?

I'm sorry, I'm not sure which toggle you mean here... I know that you implemented a toggle in your PR. As implemented in this PR, the UI will show a dropdown (not a toggle) that can add a behavior to a block. Is this what you're asking about?

<img width="290" alt="Screenshot 2023-05-15 at 15 41 39" src="https://github.com/WordPress/gutenberg/assets/5417266/6dcea6f0-d7ad-435c-a68a-360e5db43c60">


> How do you feel about adding the ability to enable or disable the lightboxes by default?

That has been discussed a little bit [here](https://github.com/WordPress/gutenberg/issues/50029#:~:text=Global%20Styles%20UI%20(not%20required%20at%20the%20beginning)). I think we can add that, but it would have to be integrated with Global Styles. It can be a follow-up PR 🙂

michalczaplinski committed:
Add a new theme for e2e tests

michalczaplinski committed:
Change theme.json to include `settings.behaviors`

michalczaplinski committed:
Remove default behavior value from block.json

michalczaplinski committed:
Minor fix to the implementation

michalczaplinski committed:
Add more e2e tests

michalczaplinski commented:
I've updated the e2e tests and added some new ones.

I've also now removed the default value from the `block.json` in https://github.com/WordPress/gutenberg/pull/49972/commits/bdcc8a5447120314298f2c141cc7fd716921e593 because otherwise the value from `settings.behaviors` would not be able to override it under any circumstances.

PS. It looks like there are some unrelated commits in the branch @c4rl0sbr4v0. Could you take a look?

I'll give it another pass tomorrow, but I think we can open it up for more review and focus on those 2 issues:

> - Figure out the right way to get the value from the top-level property of the `theme.json`
[gutenberg/packages/block-editor/src/hooks/behaviors.js](https://github.com/WordPress/gutenberg/blob/7f71175f903c7c8910e4912cd592fdaced6cfcdb/packages/block-editor/src/hooks/behaviors.js#L45-L49)
> - Make sure we consciously use `__experimentalFeatures`. It seems to be built for Global Styles, so I'm not 100% sure if our usage is correct.

cbravobernal commented:
> PS. It looks like there are some unrelated commits in the branch @c4rl0sbr4v0. Could you take a look?

It seems that when I updated with trunk, it commited instead of rebasing and force-pushing. But all commits are behaviours related, the only problem is that now I appear as commiter, although I didn't do them. Anyway, once we merge, we will squash it into only one commit, so nothing to worry about there.

I will take a look at the top-level property.

luisherranz commented on docs/reference-guides/core-blocks.md:
```diff
@@ -321,7 +321,7 @@ Insert an image to make a visual statement. ([Source](https://github.com/WordPre
 -	**Category:** media
 -	**Supports:** anchor, color (~~background~~, ~~text~~), filter (duotone)
--	**Attributes:** align, alt, caption, height, href, id, linkClass, linkDestination, linkTarget, rel, sizeSlug, title, url, width
+-	**Attributes:** align, alt, behaviors, caption, height, href, id, linkClass, linkDestination, linkTarget, rel, sizeSlug, title, url, width
```
Interestingly, the `style` attribute used by block supports is not declared (here or in the `block.json`), and still works as expected:

`<!-- wp:image { "id": 11, "style": { "border": { "width": "37px" } } } -->`

I wonder if we should treat `behaviors` in the same way and not declare it here.

michalczaplinski replied:
Yeah, that makes sense! I didn't realize it about the `style` attribute.

michalczaplinski replied:
Actually, spoke too soon! 

If we remove this from the `block.json`, then there is the following problem: 

If there is a block with the `behaviors` attribute present and in the `theme.json` we have `"behaviors": false`, then the   block's attribute will get filtered out, and the `lightbox` value will not appear in the dropdown. This is not what we want. 

To confirm this, if you remove the attribute definition from the `block.json`, the e2e test will fail on this line:

https://github.com/WordPress/gutenberg/blob/a2292f2d65491f25c2179ff7e3f04f0c8fef583c/test/e2e/specs/editor/various/behaviors.spec.js#L112

luisherranz replied:
I don't know the details of how `style` works without having to appear listed. Maybe it's whitelisted somewhere else.

luisherranz replied:
What I mean is, I don't say this should work out of the box without adding it to the attributes list, I say that `style` doesn't need to be listed and it _does_ work.

michalczaplinski replied:
Makes sense. I ll take a look at this in a follow up PR.

luisherranz commented on packages/block-library/src/image/block.json:
```diff
@@ -80,6 +80,9 @@
 			"source": "attribute",
 			"selector": "figure > a",
 			"attribute": "target"
+		},
+		"behaviors": {
+			"type": "object"
```
_See my other comment about `style` not being declared._

luisherranz commented on packages/block-editor/src/hooks/behaviors.js:
```diff
@@ -0,0 +1,96 @@
+
+		// By default, use the block behaviors.
+		// If the theme has behaviors, but the block does not, use the theme behaviors.
+		const behaviors = blockBehaviors || themeBehaviors || {};
```
Not important at this point, but this will cause problems when there is more than one behavior.

artemiomorales commented:
> I'm sorry, I'm not sure which toggle you mean here... I know that you implemented a toggle in your PR. As implemented in this PR, the UI will show a dropdown (not a toggle) that can add a behavior to a block. Is this what you're asking about?

@michalczaplinski What I meant is, should we do the two-step process like the Drop Cap currently does, making it so the dropdown enables the toggle like so (screenshotted from your video):

<img width="323" alt="Screen Shot 2023-05-16 at 4 48 03 PM" src="https://github.com/WordPress/gutenberg/assets/5360536/55ea5df6-fafc-4a28-bdef-33994dfa8cc9">

I think adding that could make for good discussion on behaviors, though it's unnecessary for the UX at the moment.

michalczaplinski commented:
> @michalczaplinski What I meant is, should we do the two-step process like the Drop Cap currently does, making it so the dropdown enables the toggle like so (screenshotted from your video):

Oh, I see! I don't think adding an extra step to show the Behaviors UI is necessary. But that's more of a UX discussion. A dropdown where you can select the behavior to be added to the block is fine. 

Also, I went for a dropdown instead of a toggle because more behaviors can likely be added to a block in the future. This way, if we add new behaviors to choose from, we wouldn't have to change the UI element.

michalczaplinski committed:
Create a selector for behaviors

michalczaplinski committed:
Add a filter to load behaviors on the server from the theme.json

michalczaplinski committed:
Add behaviors on the top-level in core theme.json

michalczaplinski committed:
Use the behaviors in the hooks

michalczaplinski committed:
Update the comment

michalczaplinski committed:
Update the comment in behaviors.js

michalczaplinski commented:
> - [ ] Figure out the right way to get the value from the top-level property of the `theme.json`
   https://github.com/WordPress/gutenberg/blob/7f71175f903c7c8910e4912cd592fdaced6cfcdb/packages/block-editor/src/hooks/behaviors.js#L45-L49
   
I've figured out the way to do that but it was non-trivial. @c4rl0sbr4v0 I explain in the video: 

<a href="https://www.loom.com/share/c80a4add5dbd4fcf8e810f203b600392">
    <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/c80a4add5dbd4fcf8e810f203b600392-with-play.gif">
  </a>
  
Overall, the whole process seems brittle to me. I would like to understand if this is the right approach. Perhaps we are OK with simply storing the values for behaviors in `settings.behaviors` which would greatly simplify the implementation.

I'd greatly appreciate your opinion here @dmsnell @gziolo @youknowriad 🙏 

Next steps if we do decide to store at the top-level:
- Figure out the way to fall back to the core `theme.json` in https://github.com/WordPress/gutenberg/blob/8b0f899f3a7ba912b5a73d439015d3e96615d97d/lib/compat/wordpress-6.3/behaviors.php#L12-L13
- Fix the e2e tests

gziolo commented:
> Overall, the whole process seems brittle to me. I would like to understand if this is the right approach. Perhaps we are OK with simply storing the values for behaviors in `settings.behaviors` which would greatly simplify the implementation.

I watched the video, and it seems like you did all steps correctly. At least, it looks similar to my experience from the past, when I had to pass additional settings to the client.

Regarding the technical aspects of how it would be best to structure everything, I believe @youknowriad and @oandregal are the best suited to share recommendations.

cbravobernal commented:
> > * [ ]  Figure out the right way to get the value from the top-level property of the `theme.json`
Thanks @michalczaplinski for the video, amazing work finding it.

What I find inconsistence is that, although the behaviors value is now in the top level, we are still calling the editor settings hook to add it. This implies that we will still need to access a "settings" (in the selector) to get that value. It seems impossible to avoid a settings selector, so, which would be the difference between accessing this new selector vs accessing the already existing settings in the theme.json?

Maybe @youknowriad or @oandregal can give us some light here, sorry for ping you again! 🙇‍♂️

cbravobernal committed:
Update fixtures and e2e

dmsnell commented on packages/block-editor/src/hooks/behaviors.js:
```diff
@@ -0,0 +1,93 @@
+						options={ Object.keys( behaviors )
+							.map( ( behavior ) => ( {
+								value: behavior,
+								label: behavior.toUpperCase(),
```
probably worth earmarking for translation issues. I'm not sure we can do this, but I'm not sure we can't either.

worth double-checking if we have other `theme.json` files that send data directly to the display like this and how they handle that.

michalczaplinski replied:
I've just realized that I've already tried this, and it's not allowed, apparently. I'm not aware of any other way to do this.

<img width="711" alt="Screenshot 2023-05-18 at 17 51 56" src="https://github.com/WordPress/gutenberg/assets/5417266/cd060b2a-9747-4008-871b-188583fe80b3">

dmsnell commented on packages/block-editor/src/hooks/behaviors.js:
```diff
@@ -0,0 +1,93 @@
+							} ) )
+							.concat( {
+								value: '',
+								label: __( 'None' ),
```
similar theme with translation: would it be helpful to specify fully what _isn't_ being selected here? something like `__( 'No behavior' )`

michalczaplinski replied:
Makes sense!

dmsnell commented on packages/block-editor/src/hooks/behaviors.js:
```diff
@@ -0,0 +1,93 @@
+							// change the default value (true) so we save it in the attributes.
+							props.setAttributes( {
+								behaviors: {
+									lightbox: nextValue === '' ? false : true,
```
not an important matter, but the ternary is redundant

```js
lightbox: nextValue !== '',
```

dmsnell replied:
or something possibly better, which is an affirmative match instead of a negative match

```js
lightbox: nextValue === 'lightbox',
```

dmsnell commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -2504,6 +2504,17 @@ export function getSettings( state ) {
 }
 
+/**
+ * Returns the behaviors.
```
I know this is a draft, but let's expand this before merging and consider what someone will learn when they find this function. what is a behavior? when should someone be using this selector?

given that the input and output types are `Object` the types give no clues to what to expect. a simple example code snippet can go a long way

```js
/**
 * Returns behaviors registered with the editor.
 *
 * "Behaviors" mark blocks for client-side interactivity in
 * predefined ways. They can be selected for blocks which have
 * declared support through `registerBlockType` and themes
 * which have done so through `theme.json`.
 *
 * Example
 *     select( 'block-editor' ).getBehaviors() === [
 *         { name: 'lightbox', … },
 *         …
 *     ]
 *
 * @see {https://behaviors-docs/somewhere}
 */
```

I'm making up the example and descriptions because I don't personally know what should go there, but this little bit of help can save someone from manually trying to find their way through the zoo of implicit dependencies in the code.

dmsnell commented on test/gutenberg-test-themes/behaviors-false/style.css:
```diff
@@ -0,0 +1,15 @@
+/*
+Theme Name: Behaviors
```
I know these are test themes and should be independent, but could it help to give them distinct names that correspond to their use? Even `Behaviors (false)` might give clues when debugging that differentiate it from the other `Behaviors` themes.

michalczaplinski commented:
> What I find inconsistence is that, although the behaviors value is now in the top level, we are still calling the editor settings hook to add it. This implies that we will still need to access a "settings" (in the selector) to get that value. It seems impossible to avoid a settings selector, so, which would be the difference between accessing this new selector vs accessing the already existing settings in the theme.json?

Hmm, I don't think it's inconsistent, but perhaps it is not clear, and we should think a bit harder about what's the best place to put the `behaviors` in the `theme.json` so that it's not confusing... 

> although the behaviors value is now in the top level, we are still calling the editor settings hook to add it.

We're using the `useSettings()` in https://github.com/WordPress/gutenberg/blob/f11ce6de0f692d7ca2efffa6e08013570872d4fe/packages/block-editor/src/hooks/behaviors.js#L35 because it's a SETTING which defines whether the Behaviors UI is shown in the Inspector Controls. It retrieves the setting in `settings.blocks.core/image.behavior.lightbox`.

> we will still need to access a "settings" (in the selector) to get that value

In terms of implementation, this is not super clear, I agree. The value for the behaviors is defined on the top-level of the `theme.json`, but then it gets shoved into the settings in:

https://github.com/WordPress/gutenberg/blob/c6dffddebce6b99872d560e047b56f3e5e644d57/packages/editor/src/components/provider/use-block-editor-settings.js#L197

https://github.com/WordPress/gutenberg/blob/c6dffddebce6b99872d560e047b56f3e5e644d57/lib/compat/wordpress-6.3/behaviors.php#L13

So we are letting users define behaviors using a top-level `theme.json` property but then shoving them into `"settings"` under the hood. 

I wonder if putting "the setting" and "the value" under this new `behaviors` key in `theme.json` would be clearer for the users. This would imply more work, though. What work? I'll quote @youknowriad from a thread in Slack:

> Are "behaviors" part of the "settings" in theme.json? We discussed previously and it seems the answer is that they shouldn't because they are more "block attribute changes" rather than "editor settings". So they're most likely a new top level key behaviors but making a new top level key have several impacts:
> - We need to update theme_json php class to properly store and validate these
> - We need to have some PHP filter in the backend that passes these behaviors in the block editor settings prop that get passed into the provider.
> - We need to make sure that key is passed between EditorProvider (outer provider) and BlockEditorProvider (inner provider)
> - We need to create a new key in the settings prop of BlockEditorProvider (we can't use __experimentalFeatures anymore)
> - We need to create a new hook to retrieve these new kind of setting useSetting is specifically tailored to retrieve from the "settings" prop.

michalczaplinski commented:
I'm stumped on the unit test error message: 

```
1) Gutenberg_REST_Block_Editor_Settings_Controller_Test::test_get_items
Undefined index: behaviors
```

@c4rl0sbr4v0 Any clue? Is it similar to the errors that you've fixed previously?

cbravobernal commented:
> @c4rl0sbr4v0 Any clue? Is it similar to the errors that you've fixed previously?

Yesterday I spent all day checking that. It seems that the `block_editor_settings_all` filter is adding a value to the response of the API that retrieves settings. So we have a new array key called behaviors that seems we need to register as a valid schema in the REST API response.

I kind of fixed (adding to `get_items` function), but then all the tests expecting response length changed to +1.

I will keep investigating and updating the PR according to.

Edit: fixed and [committed](https://github.com/WordPress/gutenberg/pull/49972/commits/201a66def019e8d4f0009732dfdeedefc344ad49).

cbravobernal committed:
Prevent mobile test gutenberg error (temporary - not the best solution)

cbravobernal committed:
Add priority on filter

cbravobernal committed:
Fix php standards

cbravobernal committed:
Found a much better way to fix php tests

cbravobernal committed:
Small refactor

michalczaplinski committed:
Add `behaviors` as an allowed key to BLOCK_EDITOR_SETTINGS

michalczaplinski committed:
Move the behaviors to top level in the e2e test theme.json file

michalczaplinski committed:
Rename the `behaviors` setting to `behaviorsUIEnabled`

michalczaplinski committed:
Change "None" to "No behaviors"

michalczaplinski committed:
Behaviors ->  behavior

michalczaplinski committed:
Fix redundant ternary

michalczaplinski committed:
Improve the JSDoc for behaviors selector

michalczaplinski committed:
Rename the test themes to make more sense

michalczaplinski commented:
Ok, I've added some minor changes. The main one is to rename the `settings.blocks[ <blockname> ].behaviors` to `behaviorsUIEnabled`. This should prevent the confusion I saw when explaining why we need two different properties in the `theme.json`.

So now, the `theme.json` could look like:

```js
// theme.json

{  
  // We can toggle the Behaviors UI on/off for specific blocks. 
  // Currently it will only be available for the core/image block.
  "settings": {
    "blocks": {
      "core/image": {
        "behaviorsUIEnabled": true
      }
    }
  },
  // Here we can define the actual behaviors that are available in the theme.
  "behaviors": {
    "lightbox": true,
   }
}
```

The immediate next steps are:
- Make sure that the Behaviors UI works in the site editor
- Make sure that the the value for `behaviors` from the core `theme.json` is being used if there is no value in the theme's `theme.json`: 
   https://github.com/WordPress/gutenberg/blob/1c39b55b18bb447166b16c702bc9898e683d7926/lib/compat/wordpress-6.3/behaviors.php#L15-L16

luisherranz commented on lib/theme.json:
```diff
@@ -1,5 +1,8 @@
 {
 	"version": 2,
+	"behaviors": {
+		"lightbox": true
+	},
```
What's the reasoning behind using `behaviors.lightbox` instead of `behaviors.blocks.core/image.lightbox`?

michalczaplinski replied:
I thought it unnecessary to nest the `behaviors` under `blocks.core/image`. I think that it's simpler and more readable without the nesting. 

Why was it considered necessary in the first place? As far as I can tell, the only use case for this namespacing would be to provide the `lightbox` behavior (or some other behaviors) as an option for one block but not for another. But we are not planning to add any other blocks with the lightbox behavior at the moment as far as I know?

luisherranz replied:
I don't know. Gallery probably. But even if we don't have plans at the moment, there's the possibility. 

`behaviors.lightbox` should be something to control _all_ the lightboxes of _all_ the blocks, not just one block. So it'd be safer to use `behaviors.blocks.core/image.lighbox` for now.

luisherranz commented on lib/theme.json:
```diff
@@ -450,6 +453,9 @@
 				}
+			},
+			"core/image": {
+				"behaviorsUIEnabled": true
```
I don't think this is consistent with the rest of the settings. None of them contain the `UIEnabled` ending.

ramonjd replied:
Welcome to the _"naming things in theme.json survivors group"_! 😄 

I take it we want to offer control over how a block "behaves", so "behavior" seems _prima facie_ like an appropriate label, assuming that future uses will fall neatly under the semantic category. 

I was wondering, which behaviors does `behaviorsUIEnabled` encompass? Is it a general flag that will be used in the future for all interactivity-related behaviors? 

So too, will there be further theme-level opt-in features that would dictate how a lightbox will display or function, e.g., animations, video support etc? 

In relation to the schema, the majority of theme.json labels describe specific, visual properties of a block, e.g., `position[ "fixed"]`, `color ["background"]` and so on. This gives us fine-grained control over which features a theme supports, and can be extended to include further properties. For this reason, I would be inclined to take inspiration from the existing block supports nomenclature.

For example, here we're disabled border radius controls for the image block, but explicitly enabling lightbox behavior.

```json
{
	"settings": {
		"blocks": {
			"core/image": {
				"border": {
					"radius": false
				},
				"behaviors": {
					"lightbox": true
				}
			},
		}
	}
}
```

To me, it makes sense to explicitly opt into a single behavior rather than blanket flag so that, as a theme developer, I'm able to fine-tune the controls available in the UI.

luisherranz replied:
That's my point. It would be weird to have to write this:

```json
{
  "settings": {
    "blocks": {
      "core/image": {
        "border": {
          "radius": false
        },
        "behaviorsUIEnabled": {
          "lightbox": true
        }
      }
    }
  }
}
```

michalczaplinski replied:
Sorry for not replying earlier; I agree and have followed your thinking on this, Luis & Ramon. Thanks for the comments!

luisherranz commented on packages/block-editor/src/hooks/behaviors.js:
```diff
@@ -0,0 +1,93 @@
+						options={ Object.keys( behaviors )
+							.map( ( behavior ) => ( {
+								value: behavior,
+								label: behavior.toUpperCase(),
+							} ) )
+							.concat( {
+								value: '',
+								label: __( 'No behaviors' ),
```
Why are the behavior names uppercase in the UI? Is it needed? Maybe we can define their UI name somewhere.

michalczaplinski replied:
There's no specific reason other than that I thought it looked better 🙁. Perhaps they could just be capitalized like the names of the specific image sizes:

<img width="269" alt="Screenshot 2023-05-22 at 15 48 51" src="https://github.com/WordPress/gutenberg/assets/5417266/d2756c00-0cb1-4f41-8933-c9095a04c8b7">


<img width="265" alt="Screenshot 2023-05-22 at 15 46 00" src="https://github.com/WordPress/gutenberg/assets/5417266/5854babe-2fde-4cf4-a8e5-3ad657eea1bc">

michalczaplinski replied:
OK, I've capitalized the behavior names now. I think it's more consistent this way.

luisherranz reviewed (commented):
I'm glad to see this is finally coming along. Good work 🙂

I have a few comments, added below.

dmsnell commented on lib/compat/wordpress-6.3/behaviors.php:
```diff
@@ -0,0 +1,19 @@
+		// TODO: Make sure to also get the value from the core theme.json file.
+		return $settings;
+	},
+	PHP_INT_MAX
```
is this for development only? if so maybe let's add a `@TODO` to remove this before merge.

michalczaplinski replied:
yes, this is still a TODO to be done before merging 👍

michalczaplinski replied:
FYI, I've now moved the PHP code specific to the behaviors to [`block-editor-settings.php`](https://github.com/WordPress/gutenberg/blob/fc812bc6c04e8b3c5c7fe3450c1e39d1dd6f6f3b/lib/block-editor-settings.php)

michalczaplinski replied:
FYI 2: I've reverted the change from above because I've realized that `block-editor-settings.php` is for GB-plugin-specific stuff only. Sorry for the complication 🙂

gziolo replied:
The current priority prevents other plugins from updating the `behaviors` fields in the settings passed to the client.

michalczaplinski replied:
I see, thanks for pointing this out Greg. Will fix it!

michalczaplinski replied:
Just to clarify, will fix it in a follow-up PR most probably 👍

michalczaplinski committed:
Remove definition of `behaviors` attribute in core/image

michalczaplinski committed:
Change default value for `behaviors.lightbox` to false and update e2e tests accordingly

michalczaplinski committed:
Change the way we get the data from `theme.json` and adjust e2e

michalczaplinski committed:
Capitalize behaviors' labels

ramonjd commented on test/gutenberg-test-themes/behaviors-enabled/theme.json:
```diff
@@ -0,0 +1,6 @@
+{
+	"version": 2,
+	"behaviors": {
+		"lightbox": true
```
I think it's fine that the labels match the ones we use in blocks. It follows an existing pattern (see block supports) and will probably be easier for folks to grok.

ramonjd commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -404,6 +405,7 @@ class WP_Theme_JSON_Gutenberg {
 			'textDecoration' => null,
 			'textTransform'  => null,
 		),
+		'behaviorsUIEnabled'            => null,
```
Same as [below](https://github.com/WordPress/gutenberg/pull/49972/files#r1201361853).

My instinct is to be explicit and modular here. Do we assume that theme builders will understand any implied semantics in `-UIEnabled`?

Rather, an independent unit `behavior.lightbox` tells us what we're dealing with immediately. 

If the term "behavior" doesn't have the semantic punch to convey what the sub properties are doing, then it's okay to stop and spend time thinking about it. It's far better than doing the backwards compatibility cha-cha down the road. 😄

michalczaplinski committed:
Move PHP code adding `theme.json` behaviors to `block-editor-settings`

michalczaplinski committed:
Update comment

michalczaplinski committed:
Remove the behaviors require from load.php

michalczaplinski committed:
Revert "Update comment"

This reverts commit fc812bc6c04e8b3c5c7fe3450c1e39d1dd6f6f3b.

michalczaplinski committed:
Revert "Move PHP code adding `theme.json` behaviors to `block-editor-settings`"

This reverts commit e8d16c8bb4977d1e27b758a067066a78fc21f794.

michalczaplinski commented:
I've pushed a few more changes to iron out the issues mentioned by @luisherranz and other improvements:
- Capitalize the labels of the behaviors in the UI (`Lightbox` instead of `LIGHTBOX` in all caps)
- [Change default value for behaviors.lightbox to false and update e2e tests](https://github.com/WordPress/gutenberg/pull/49972/commits/697c04fde2a1710abce8a72cf9bb3fd99d0849f6)
- [Change the way we get the data from theme.json and adjust e2e](https://github.com/WordPress/gutenberg/pull/49972/commits/a2292f2d65491f25c2179ff7e3f04f0c8fef583c) because previously the core `theme.json` file was being ignored.

The last thing to do before merging is to decide on the naming:
  - `behaviors.lightbox` vs. `behaviors.blocks.<block-name>.lightbox`
  - `settings.blocks.<block-name>.behaviors` vs `settings.blocks.<block-name>.behaviorsUIEnabled`
  
As for the site editor, it doesn't work there just yet. We can figure it out in a quick follow up PR though.

youknowriad commented:
> behaviors.lightbox vs. behaviors.blocks.<block-name>.lightbox

This feels like we could support both. One enabling light boxes for all blocks, and one for a particular block

> settings.blocks.<block-name>.behaviors vs settings.blocks.<block-name>.behaviorsUIEnabled

the former for me for consistency with all of our settings, I'd also add that maybe `settings.blocks.<block-name>.behaviors.lightbox` is better in case we add new behaviors and we want to show some of them selectively in the UI.

michalczaplinski committed:
Remove the comment that was added previously

michalczaplinski committed:
Update comments in `behaviors.php`

michalczaplinski committed:
Add back the require_once in load.php

michalczaplinski committed:
Use `settings.blocks.core/image.behaviors.lightbox`

michalczaplinski committed:
Use `behaviors.blocks.core/image.lightbox`

michalczaplinski commented:
@youknowriad 

> This feels like we could support both. One enabling light boxes for all blocks, and one for a particular block

gotcha, that makes sense. For now, we'll only gonna support `behaviors.blocks.<block-name>.lightbox` but we can extend it in the future.

michalczaplinski commented:
⚠️ **IMPORTANT**: Before merging this, you should change the base of the https://github.com/WordPress/gutenberg/pull/50373 PR back to `trunk` so that PRs history does not get lost. Apparently it's a github bug.


I've updated the naming to match the consensus:

- `behaviors.blocks.<block-name>.lightbox` for the values of the behaviors.
- `settings.blocks.<block-name>.behaviors` for the settings.

The tests are passing. I'd appreciate a final review so that we can push it over the finish line asap @c4rl0sbr4v0 @luisherranz @artemiomorales @SantosGuillamot 

Making it work correctly in the site editor will be addressed in follow-up PR.

youknowriad commented on packages/block-editor/src/hooks/behaviors.js:
```diff
@@ -0,0 +1,104 @@
+			Object.entries( settings ).every( ( [ , value ] ) => ! value )
+		) {
+			return <BlockEdit { ...props } />;
+		}
```
There's a subtile issue that may or may not result in a bug. The fact that we conditionally render `BlockEdit` can result in a "remount" of the whole component if the condition above changes from "true" to "false" or the opposite. 

The remount often create focus loss and issues like that. While this might not always be a problem, it is a problem in general if the condition "can" change when the component is mounted (I'm not sure it's the case here)

cbravobernal replied:
That conditional is reading a theme.json value. I don't think that condition can change without a page reload 🤔

youknowriad replied:
Yeah, probably fine for now, just something to know as sometimes we used to switch over attribute values (easy to add a check there). Also, maybe at some point in the future, that setting could be editable in the global styles UI in the site editor.

cbravobernal replied:
> that setting could be editable in the global styles UI in the site editor

That is the next step after merging this PR, so we should then take a look at this issue.

cbravobernal reviewed (approved):
Looks good to me. I won't enable auto-merge as we need to update another PR as requested in a comment.

SantosGuillamot reviewed (approved):
I've been testing the UX and it looks great 👏 Some comments:

### Testing

I used:
* `settings.blocks.core/image.behaviors.lightbox`: To enable/disable the Editor UI to select behaviors per block.
* `behaviors.blocks.core/image.lightbox`: To define a default value and activate/deactivate the lightbox in the images.

What I tested:

1. In the theme `theme.json`: NO `settings.blocks.core/image.behaviors.lightbox` and NO `behaviors.blocks.core/image.lightbox`

It gets the default values from Gutenberg and it shows the UI to select the behaviors and the default value is "No behaviors".

2. In the theme `theme.json`: Add `settings.blocks.core/image.behaviors.lightbox = false`

The UI to select the behaviors doesn't appear anymore.

3. In the theme `theme.json`: Add `behaviors.blocks.core/image.lightbox = true` and remove the `settings.blocks.core/image.behaviors.lightbox`

The UI appears and the default value is "Lightbox". It works in the frontend.

4. In the theme `theme.json`: Add `settings.blocks.core/image.behaviors.lightbox = false` and `behaviors.blocks.core/image.lightbox = true`

The UI to select the behaviors doesn't appear in the Block Editor BUT the lightbox is still working in the frontend.

5. Overwrite the value for each block.

If I select a behavior from the block, it overwrites the default value defined by the theme or Gutenberg.

### Potential improvements

I just have a few aspects to consider:

* I think **we might need a "Reset" button** to remove the lightbox from the block attributes and listen to the default value from the theme/Gutenberg. At this moment, if I select "Lightbox" from the Editor UI, there is no way to go back to the default value. I can change the attribute to "No behaviors", but it is not the same as listening to the default. This is something many styles are using for example.
* We might need to **add the behaviors property to the [schemas](https://github.com/WordPress/gutenberg/tree/trunk/schemas/json)**.
* I didn't test it, but I guess that, in the future when there are more behaviors, if `settings.blocks.core/image.behaviors.lightbox` is false, **it shouldn't appear in the dropdown list but it shouldn't remove the whole list of behaviors**.

cbravobernal committed:
Remove experimental setting for interactivity API

cbravobernal committed:
Revert "Remove experimental setting for interactivity API"

This reverts commit 70e51cd684d186668eea9ff9c3786f38209ffcd5.

SantosGuillamot committed:
Move `interactivity.js` files to `block.json`

SantosGuillamot committed:
Remove experimental flag for the Interactivity API

kevin940726 commented on test/e2e/specs/editor/various/behaviors.spec.js:
```diff
@@ -0,0 +1,161 @@
+const { test, expect } = require( '@wordpress/e2e-test-utils-playwright' );
+
+test.describe( 'Testing behaviors functionality', () => {
+	const filename = '1024x768_e2e_test_image_size.jpeg';
+	const filepath = path.join( './test/e2e/assets', filename );
+
+	const createMedia = async ( { admin, requestUtils } ) => {
+		await admin.createNewPost();
+		const media = await requestUtils.uploadMedia( filepath );
+		return media;
+	};
```
It's a convention to use the [POM](https://playwright.dev/docs/pom)-style class as shown in other test file to create file-specific utilities. See the `image.spec.js` test for an example. It's not strictly required though, it's just a suggestion :)!

kevin940726 commented on test/e2e/specs/editor/various/behaviors.spec.js:
```diff
@@ -0,0 +1,161 @@
+			},
+		} );
+
+		await page.getByRole( 'button', { name: 'Advanced' } ).click();
```
Nit: We can scope it under the `Editor settings` sidebar to make it more readable. (It isn't clear which "Advanced" button we're clicking as it's a pretty generic name.)

```suggestion
		await page
			.getByRole( 'region', { name: 'Editor settings' } )
			.getByRole( 'button', { name: 'Advanced' } )
			.click();
```

kevin940726 commented on test/e2e/specs/editor/various/behaviors.spec.js:
```diff
@@ -0,0 +1,161 @@
+		} );
+
+		await page.getByRole( 'button', { name: 'Advanced' } ).click();
+		const select = page.getByLabel( 'Behavior' );
```
Try `getByRole` with parent locator here too?

```suggestion
		const select = editorSettings.getByRole( 'combobox', { name: 'Behavior' } );
```

kevin940726 commented on test/e2e/specs/editor/various/behaviors.spec.js:
```diff
@@ -0,0 +1,161 @@
+		const select = page.getByLabel( 'Behavior' );
+
+		// By default, no behaviors should be selected.
+		await expect( select ).toHaveCount( 1 );
```
This isn't needed as a locator is strict by default.

kevin940726 commented on test/e2e/specs/editor/various/behaviors.spec.js:
```diff
@@ -0,0 +1,161 @@
+		await expect( select ).toHaveValue( '' );
+
+		// By default, you should be able to select the Lightbox behavior.
+		const options = select.locator( 'option' );
```
Ditto. Try `getByRole` instead?

```suggestion
		const options = select.getByRole( 'option' );
```

kevin940726 commented on test/e2e/specs/editor/various/behaviors.spec.js:
```diff
@@ -0,0 +1,161 @@
+		await page.getByRole( 'button', { name: 'Advanced' } ).click();
+
+		// No behaviors dropdown should be present.
+		await expect( page.getByLabel( 'Behavior' ) ).toHaveCount( 0 );
```
A more readable and explicit assertion might be:

```suggestion
		await expect( locator ).toBeHidden();
```

michalczaplinski commented:
@SantosGuillamot Thank you for testing thorougly 🙏 

> - I think we might need a "Reset" button to remove the lightbox from the block attributes and listen to the default value from the theme/Gutenberg. At this moment, if I select "Lightbox" from the Editor UI, there is no way to go back to the default value. I can change the attribute to "No behaviors", but it is not the same as listening to the default. This is something many styles are using for example.
> - We might need to add the behaviors property to the [schemas](https://github.com/WordPress/gutenberg/tree/trunk/schemas/json?rgh-link-date=2023-05-24T09%3A20%3A30Z).
> - I didn't test it, but I guess that, in the future when there are more behaviors, if settings.blocks.core/image.behaviors.lightbox is false, it shouldn't appear in the dropdown list but it shouldn't remove the whole list of behaviors.

Agreed on all 3 points. I ll work on this in a follow up PR.

michalczaplinski committed:
Revert "Move `interactivity.js` files to `block.json`"

This reverts commit c32be86e150ccd8213167287b473774c6f80d6bc.

michalczaplinski committed:
Revert "Remove experimental flag for the Interactivity API"

This reverts commit b8dc7dd6ff6bec3b9a4576dd8341f27fca5b9dfb.

michalczaplinski commented:
@kevin940726 Thanks for your comments! @Mamaduka has [implemented](https://github.com/WordPress/gutenberg/pull/50954) the improvements you suggested, and I've just merged them 🙂 .

michalczaplinski commented:
I've filed the backport for WP 6.3 here: https://github.com/WordPress/wordpress-develop/pull/4526
